File: app/src/main/java/xyz/klinker/messenger/service/MediaParserService.java
package xyz.klinker.messenger.service;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;

import xyz.klinker.messenger.data.DataSource;
import xyz.klinker.messenger.data.model.Message;
import xyz.klinker.messenger.receiver.MessageListUpdatedReceiver;
import xyz.klinker.messenger.util.media.MediaMessageParserFactory;
import xyz.klinker.messenger.util.media.MediaParser;

public class MediaParserService extends IntentService {

    public static final String EXTRA_CONVERSATION_ID = "conversation_id";
    public static final String EXTRA_BODY_TEXT = "body_text";

    public MediaParserService() {
        super("MediaParserService");
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        long conversationId = intent.getLongExtra(EXTRA_CONVERSATION_ID, -1L);
        String text = intent.getStringExtra(EXTRA_BODY_TEXT);

        if (conversationId == -1L || text == null) {
            return;
        }

        MediaParser parser = new MediaMessageParserFactory().getInstance(this, text);

        if (parser == null) {
            return;
        }

        Message message = parser.parse(conversationId);

        DataSource source = DataSource.getInstance(this);
        source.open();
        source.insertMessage(this, message, conversationId, true);
        source.close();

        MessageListUpdatedReceiver.sendBroadcast(this, conversationId, message.data, message.type);
    }
}

File: shared/src/main/java/xyz/klinker/messenger/shared/service/message_parser/MediaParserService.kt
package xyz.klinker.messenger.shared.service.message_parser

import android.annotation.SuppressLint
import android.app.IntentService
import android.content.Context
import android.content.Intent
import android.support.v4.app.NotificationCompat
import xyz.klinker.messenger.shared.R
import xyz.klinker.messenger.shared.data.ColorSet
import xyz.klinker.messenger.shared.data.DataSource
import xyz.klinker.messenger.shared.data.Settings
import xyz.klinker.messenger.shared.data.model.Message
import xyz.klinker.messenger.shared.receiver.MessageListUpdatedReceiver
import xyz.klinker.messenger.shared.util.AndroidVersionUtil
import xyz.klinker.messenger.shared.util.NotificationUtils
import xyz.klinker.messenger.shared.util.media.MediaMessageParserFactory
import xyz.klinker.messenger.shared.util.media.MediaParser
import xyz.klinker.messenger.shared.util.media.parsers.ArticleParser

class MediaParserService : IntentService("MediaParserService") {

    override fun onHandleIntent(intent: Intent?) {
        if (AndroidVersionUtil.isAndroidO) {
            val notification = NotificationCompat.Builder(this,
                    NotificationUtils.SILENT_BACKGROUND_CHANNEL_ID)
                    .setContentTitle(getString(R.string.media_parse_text))
                    .setSmallIcon(R.drawable.ic_stat_notify_group)
                    .setProgress(0, 0, true)
                    .setLocalOnly(true)
                    .setColor(ColorSet.DEFAULT(this).color)
                    .setOngoing(true)
                    .setPriority(NotificationCompat.PRIORITY_MIN)
                    .build()
            startForeground(MEDIA_PARSE_FOREGROUND_ID, notification)
        }

        if (intent == null) {
            stopForeground(true)
            return
        }

        val messageId = intent.getLongExtra(EXTRA_MESSAGE_ID, -1L)
        val message = DataSource.getMessage(this, messageId)

        if (message == null) {
            stopForeground(true)
            return
        }

        val parser = createParser(this, message)
        if (parser == null || (!Settings.internalBrowser && parser is ArticleParser)) {
            stopForeground(true)
            return
        }

        val parsedMessage = parser.parse(message)
        if (parsedMessage != null) {
            DataSource.insertMessage(this, parsedMessage, message.conversationId, true)
            MessageListUpdatedReceiver.sendBroadcast(this, message.conversationId, parsedMessage.data, parsedMessage.type)
        }

        if (AndroidVersionUtil.isAndroidO) {
            stopForeground(true)
        }
    }

    companion object {

        @SuppressLint("NewApi")
        fun start(context: Context, message: Message) {
            val mediaParser = Intent(context, MediaParserService::class.java)
            mediaParser.putExtra(EXTRA_MESSAGE_ID, message.id)

            if (AndroidVersionUtil.isAndroidO) {
                context.startForegroundService(mediaParser)
            } else {
                context.startService(mediaParser)
            }
        }

        private const val MEDIA_PARSE_FOREGROUND_ID = 1334

        const val EXTRA_MESSAGE_ID = "message_id"

        fun createParser(context: Context, message: Message): MediaParser? {
            return MediaMessageParserFactory().getInstance(context, message)
        }
    }
}
